magicgui/_magicgui.py

```diff
@@ -1,18 +1,70 @@
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Callable, Optional, Union, overload
+import inspect
+from functools import partial
+from types import FunctionType
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Callable,
+    Generic,
+    Optional,
+    TypeVar,
+    Union,
+    overload,
+)
 from warnings import warn
 
 from typing_extensions import Literal
 
+from magicgui.widgets import FunctionGui, MainFunctionGui
+
 if TYPE_CHECKING:
     from magicgui.application import AppRef
-    from magicgui.widgets import FunctionGui, MainFunctionGui
+
+_T = TypeVar("_T")
+_R = TypeVar("_R")
+
+
+def _magicgui(function=None, factory=False, main_window=False, **kwargs):
+    """Actual private magicui decorator.
+
+    if factory is `True` will return a MagicFactory instance, that can be called
+    to return a `FunctionGui` instance.  See docstring of ``magicgui`` for parameters
+    """
+    if "result" in kwargs["param_options"]:
+        warn(
+            "\n\nThe 'result' option is deprecated and will be removed in the future."
+            "Please use `result_widget=True` instead.\n",
+            FutureWarning,
+        )
+
+        kwargs["param_options"].pop("result")
+        kwargs["result_widget"] = True
```

Comment:
what did `result` do before? I'm confused about why this isn't `kwargs["result_widget"] = kwargs["param_options"].pop("result")`

Reply:
Oh, was "result" a special parameter name in magicgui? That's pretty terrible. LOL I might consider removing it altogether immediately. (But probably in a separate PR.) Generally, special names like that should probably have underscores or other prefixes so they don't conflict with actual parameters.

Reply:
yeah that seems very reasonable!  I need to go back and look now, but it's possible it used to accept a string?

Reply:
yeah, it was terrible! I agree!  I can remove in another PR

Reply:
#122